Migrate wrong datasource UIDs

andresmgot opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

This PR:
 - Adds logging when trying to get a data source with an invalid UID.
 - Migrates datasources with an invalid UID (behind a feature flag).

Note that enabling the feature flag would break any dashboard / alert / other query referencing the old (invalid UID).

The goal is to get more information to check if there are data sources with wrong UIDs in actual use. If data sources are not in use, they can be migrated without further issues (enabling the feature flag).

**Why do we need this feature?**


**Who is this feature for?**


**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Ref https://github.com/grafana/grafana/issues/73892

**Special notes for your reviewer:**

Things that I tried and discarded:

 - Automatically migrate datasources with [datasource_mig.go](https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/migrations/datasource_mig.go). There is not a good way to do a regex replace for the unsupported characters. We could specify some characters (like empty spaces or `/`) but we could miss some cases.
 - Automatically migrate dashboards with [DashboardMigrator.ts](https://github.com/grafana/grafana/blob/main/public/app/features/dashboard/state/DashboardMigrator.ts). Two problems here: One is that it requires all datasources to be migrated first (so the point above would need to be solved). The other is that there are multiple places in which a datasource may be referenced (queries, annotations, variables), which make it a bit tricky to cover everything.
 - Automatically migrate dashboards and alerts with `alert_mig.go` and `dashboard_mig.go`. Apart than the problems already described, the queries are encoded as JSON blobs (`data` column), making it more difficult to make replacements.
 - Automatically return a data source with a valid UID when requesting an invalid one. As @marefr pointed out, this could lead to an escalation of privileges when a request to a protected datasource is made using permissions of the invalid UID since that's the one being checked.

Please check that:
- [x] It works as expected from a user's perspective.
- [x] If this is a pre-GA feature, it is behind a feature toggle.
- [x] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

andresmgot committed:
Migrate wrong datasource UIDs

andresmgot committed:
Add console.warn

andresmgot committed:
Allow to use wrong UIDs

andresmgot committed:
replace warn

marefr reviewed (commented):
Sorry I'm a bit confused so need to ask some additional questions.

> Provisioned datasources will be automatically migrated to a valid UID.

How? Is this because provisioning looks up datasources by name?

> Automatically migrate datasources with [datasource_mig.go](https://github.com/grafana/grafana/blob/main/pkg/services/sqlstore/migrations/datasource_mig.go). There is not a good way to do a regex replace for the unsupported characters. We could specify some characters (like empty spaces or /) but we could miss some cases.

There's also code migrations as an alternative, but understand it might be tricky in general. 

This doesn't address alert rules or does it?

My personal opinion is returning an error wherever an invalid uid is referenced because
- We have deprecated the usage
- Should be up to the consumer to change/decide the new uid
- Might be other places datasource uid's are used, such as the correlations feature perhaps? Maybe that's already works with this solution?

Ryan suggested earlier that introducing a new column (or table) to hold old uid's to support fallback looking up by the old. With that we could keep the existing ones, but could also store the generated/valid ones. Might be worth exploring perhaps?

andresmgot commented:
> > Provisioned datasources will be automatically migrated to a valid UID.
> 
> How? Is this because provisioning looks up datasources by name?
> 

By `id` and `org_id` :shrug: [see the code](https://github.com/grafana/grafana/blob/wrongUIDs/pkg/services/datasources/service/store.go#L397)

> This doesn't address alert rules or does it?

It does because when evaluating an alert, it retrieves the data source config with `GetDatasource` and I have added the [fallback code to fix UIDs on the fly](https://github.com/grafana/grafana/blob/wrongUIDs/pkg/services/datasources/service/store.go#L81-L93).

> My personal opinion is returning an error wherever an invalid uid is referenced because
> 
> * We have deprecated the usage
> * Should be up to the consumer to change/decide the new uid
> * Might be other places datasource uid's are used, such as the correlations feature perhaps? Maybe that's already works with this solution?

The idea of this PR is to avoid the breaking change, we would need to help with the migration in Cloud anyway so this could be a solution that fits all the use cases.

Regarding testing the solution with other features, I am happy to test any other scenario. In theory, this changes should fix its usage for both frontend and backend features but I don't know all the possible use cases.

> Ryan suggested earlier that introducing a new column (or table) to hold old uid's to support fallback looking up by the old. With that we could keep the existing ones, but could also store the generated/valid ones. Might be worth exploring perhaps?

I think that adding another UID field may bit us in unexpected ways if we try to use it in functional code. If you mean to just store it so it would easily allow us to revert the changes if needed, that could be an option. I don't know if we need to worry about leaving "garbage" behind though.

andresmgot committed:
fixes

andresmgot committed:
fix regex

andresmgot committed:
Merge branch 'main' into wrongUIDs

andresmgot committed:
more fixes

marefr commented:
> By `id` and `org_id` 🤷 [see the code](https://github.com/grafana/grafana/blob/wrongUIDs/pkg/services/datasources/service/store.go#L397)

The id needs to be looked up by `org_id` and `name` in provisioning:
https://github.com/grafana/grafana/blob/12771e49fc13b4b6e7e073100c4f7d0f91d915a5/pkg/services/provisioning/datasources/datasources.go#L67

> It does because when evaluating an alert, it retrieves the data source config with `GetDatasource` and I have added the [fallback code to fix UIDs on the fly](https://github.com/grafana/grafana/blob/wrongUIDs/pkg/services/datasources/service/store.go#L81-L93).

> The idea of this PR is to avoid the breaking change, we would need to help with the migration in Cloud anyway so this could be a solution that fits all the use cases.

Thanks I understand now 👍 Was a bit too fast reviewing this end of last week. Let me have another look

marefr commented on pkg/services/datasources/service/store.go:
```diff
@@ -77,6 +73,28 @@ func (ss *SqlStore) getDataSource(ctx context.Context, query *datasources.GetDat
+		return nil, datasources.ErrDataSourceIdentifierNotSet
+	}
+
+	if err := util.ValidateUID(query.UID); len(query.UID) > 0 && err != nil {
```
I wonder if we should optimize for a valid UID and query with the provided UID first. You even mention below "I assume there won't many". Then if no match, validate, fix and re-query using the fixed uid?

After consideration, I see one potential (security) flaw with this approach and that is that this could be used for someone to craft a datasource with a uid to get access to a datasource they don't have access to which is not good. To secure this we probably need to implement this higher up where permission checks happens 😢

andresmgot replied:
> I wonder if we should optimize for a valid UID and query with the provided UID first. You even mention below "I assume there won't many". Then if no match, validate, fix and re-query using the fixed uid?

Good point.

> After consideration, I see one potential (security) flaw with this approach and that is that this could be used for someone to craft a datasource with a uid to get access to a datasource they don't have access to which is not good. To secure this we probably need to implement this higher up where permission checks happens 😢

True. It's a bit of a corner case because the user would need to have specific access for the wrong UID (e.g. `custom/uid`) but not for the valid representation (`custom-uid`) plus it would require that the datasource with the wrong UID doesn't exist (so we fallback to check the valid UID as per your comment above).

In any case, we may not need this if the datasources with wrong UIDs are not used (since they are just pre-created). I am going to put this behind a feature flag, keeping the logging so we can get more information.

marefr replied:
>  it would require that the datasource with the wrong UID doesn't exist 

Yes. So even if we did your original validation/fix or my proposal they both has the same flaw

> In any case, we may not need this if the datasources with wrong UIDs are not used (since they are just pre-created)

Think you can expand on this, pre-created especially?

andresmgot replied:
> Think you can expand on this, pre-created especially?

I just meant automatically provisioned, without user interaction.

andresmgot committed:
Merge branch 'main' into wrongUIDs

andresmgot committed:
Move to feature flag. Revert read changes

andresmgot committed:
clean up changes

andresmgot committed:
fix test

andresmgot committed:
Merge branch 'main' into wrongUIDs

andresmgot commented:
@marefr reduced the scope of this, see the new PR description.

andresmgot committed:
Merge branch 'main' into wrongUIDs

marefr reviewed (commented):
Thanks. 

> The goal is to get more information to check if there are data sources with wrong UIDs in actual use. If data sources are not in use, they can be migrated without further issues (enabling the feature flag).

Still a bit confused. Thought we already identified there was just a few invalid ones in Grafana Cloud? Or did that investigation not cover everything?

In addition, if there are no issues identified I would suggest that the feature toggle enabled validation and failure to stop the possibility of new invalid uid's being set rather than automatically fixing them. Thoughts?

andresmgot commented:
> Still a bit confused. Thought we already identified there was just a few invalid ones in Grafana Cloud? Or did that investigation not cover everything?

The log warning that we checked `"Invalid datasource uid... ` is only printed when creating or updating a data source. The log line line is exactly the same for the update and the creation case. What I have changed is that the log line now also prints the action (`update`, `create`) and I have also added the warning for the case in which the data source is accessed (`read`). This allows us to know if the log line is only appearing because the same provisioning file is still being used (we will only see `update` entries) or if it is also used in dashboards or alerts (we will get `read` entries).

> In addition, if there are no issues identified I would suggest that the feature toggle enabled validation and failure to stop the possibility of new invalid uid's being set rather than automatically fixing them. Thoughts?

For the case of creating new data sources I agree but I wanted to double check first that there is nothing creating datasources with wrong UIDs to avoid a breaking change there. For the update case, it's less disruptive if we automatically fix those. Does it make sense?

andresmgot committed:
Merge branch 'main' into wrongUIDs

andresmgot committed:
fail on datasource creation

andresmgot committed:
merge main

marefr commented:
> The log warning that we checked "Invalid datasource uid...  is only printed when creating or updating a data source. The log line line is exactly the same for the update and the creation case. What I have changed is that the log line now also prints the action (update, create) and I have also added the warning for the case in which the data source is accessed (read). This allows us to know if the log line is only appearing because the same provisioning file is still being used (we will only see update entries) or if it is also used in dashboards or alerts (we will get read entries).

Got it. Thought that was extracted without looking at logs.

> For the case of creating new data sources I agree but I wanted to double check first that there is nothing creating datasources with wrong UIDs to avoid a breaking change there. For the update case, it's less disruptive if we automatically fix those. Does it make sense?

Yes, but perhaps skip the feature toggle/auto fix for now if we're only intending capturing more logs?

andresmgot commented:
> Yes, but perhaps skip the feature toggle/auto fix for now if we're only intending capturing more logs and we're not sure that autofix or validation error will be the outcome?

The validation error is what we want to avoid, so it's not a breaking change. My idea was to potentially enable the feature flag for instances that may need it (that has datasources with invaliud UIDs but are not in use) and do the migration that way. Thoughts @tolzhabayev ?

marefr commented:
> The validation error is what we want to avoid, so it's not a breaking change. 
Okay sure

> My idea was to potentially enable the feature flag for instances that may need it (instances that have datasources with invaliud UIDs but are not in use) and do the migration that way. Thoughts @tolzhabayev ?

How do we know of instances that have datasources with invaliud UIDs but are not in use?

andresmgot commented:
> How do we know of instances that have datasources with invaliud UIDs but are not in use?

Those will show an "update" event (because the data source is part of its provisioning) but not a "read" event, which happens when the data source is used.

marefr reviewed (approved):
LGTM lets get the improved logging out

tolzhabayev commented:
@andresmgot a bit late to the party as it is merged - but from my understanding the latest state of the PR (I did not follow all the commits, just what it is right now) would auto fix UID on `UpdateDataSource` silently and error out on `AddDataSource` when called with a wrong UID. All of that when the feature flag is enabled.

Meaning we probably won't be able to migrate that way (as update still needs to be called and it might happen or not - it's not really something we control) but we will be able to collect more logs to see how they are affected and prevent creating new datasource instances with wrong uids - so lgtm let's see what logs give us.
